crates/core/src/node/inner/vm_runner.rs

```diff
@@ -119,46 +120,69 @@ impl VmRunner {
+                    transaction_hash: Box::new(tx_hash),
+                    tx_gas_per_pubdata_limit: Box::new(tx_data.fee.gas_per_pubdata_limit),
+                    max_gas: Box::new(max_gas),
+                },
+            );
         }
 
         let l2_gas_price = batch_env.fee_input.fair_l2_gas_price();
         if tx_data.fee.max_fee_per_gas < l2_gas_price.into() {
-            sh_eprintln!(
-                "Submitted Tx is Unexecutable {:?} because of MaxFeePerGasTooLow {}",
-                tx_hash,
-                tx_data.fee.max_fee_per_gas
+            return Err(
+                anvil_zksync::node::TransactionValidationFailedMaxFeePerGasTooLow {
+                    max_fee_per_gas: Box::new(tx_data.fee.max_fee_per_gas),
+                    transaction_hash: Box::new(tx_hash),
+                    l2_gas_price: Box::new(l2_gas_price.into()),
+                },
             );
-            anyhow::bail!("block base fee higher than max fee per gas");
         }
 
         if tx_data.fee.max_fee_per_gas < tx_data.fee.max_priority_fee_per_gas {
-            sh_eprintln!(
-                "Submitted Tx is Unexecutable {:?} because of MaxPriorityFeeGreaterThanMaxFee {}",
-                tx_hash,
-                tx_data.fee.max_fee_per_gas
+            return Err(
+                anvil_zksync::node::TransactionValidationFailedMaxPriorityFeeGreaterThanMaxFee {
+                    max_fee_per_gas: Box::new(tx_data.fee.max_fee_per_gas),
+                    max_priority_fee_per_gas: Box::new(tx_data.fee.max_priority_fee_per_gas),
+                    transaction_hash: Box::new(tx_hash),
+                },
             );
-            anyhow::bail!("max priority fee per gas higher than max fee per gas");
         }
         Ok(())
     }
 
     async fn run_tx_pretty(
         &mut self,
-        tx: Transaction,
+        tx: &Transaction,
         executor: &mut dyn BatchExecutor<StorageView<ForkStorage>>,
         config: &TestNodeConfig,
         fee_input_provider: &TestNodeFeeInputProvider,
-    ) -> anyhow::Result<BatchTransactionExecutionResult> {
+    ) -> Result<BatchTransactionExecutionResult, AnvilNodeError> {
         let BatchTransactionExecutionResult {
             tx_result,
             compression_result,
             call_traces,
         } = executor.execute_tx(tx.clone()).await?;
-        compression_result?;
+        compression_result.map_err(|_inner|
+                                   // We ignore `inner` because bytecode
+                                   // compression error currently does not hold
+                                   // any precise information
+                                   anvil_zksync::node::TransactionHalt {
+                                       inner: Box::new(anvil_zksync::halt::FailedToPublishCompressedBytecodes),
+                                       transaction_hash: Box::new(tx.hash()),
+                                   })?;
```

Comment:
Nit: this is kind of ugly because of the comment block, do you mind wrapping everything in `{}`?